File: src/frontend/src/api/aiTrading.ts
/**
 * AI Trading API client.
 *
 * Provides methods for natural language trading execution,
 * trade confirmation, configuration, and history.
 */

import request from './index'

const API_BASE = '/ai-trading'

export interface TradingIntent {
  action: string
  symbol: string | null
  exchange: string | null
  quantity: number | null
  price: number | null
  order_type: string
  stop_loss: number | null
  take_profit: number | null
  reason: string
  confidence: number
  risk_level: string
  raw_input: string
}

export interface RiskAssessment {
  approved: boolean
  risk_level: string
  warnings: string[]
  blocked_reasons: string[]
  requires_confirmation: boolean
  max_loss_estimate: number | null
  position_impact: string | null
}

export interface AITradingGatewayOption {
  gateway_id: string
  exchange_type: string
  account_id: string
  connected: boolean
}

export interface AITradingAccountOption {
  account_id: string
  name: string
  total_equity?: number | null
  current_cash?: number | null
  is_active: boolean
  source?: 'paper' | 'gateway' | string
  gateway_id?: string | null
  exchange_type?: string | null
  connected?: boolean
}

export interface AITradingResponse {
  trade_id: string
  intent: TradingIntent
  risk_assessment: RiskAssessment
  status: string
  message: string
  execution_result: Record<string, unknown> | null
  ai_reasoning: string
  suggestions: string[]
  requires_confirmation: boolean
  degraded: boolean
  diagnostic_message: string | null
}

export interface AITradingConfig {
  enabled: boolean
  default_mode: string
  max_single_trade_amount: number
  max_daily_trades: number
  max_position_ratio: number
  require_confirmation_above: number
  blocked_symbols: string[]
  available_gateways: AITradingGatewayOption[]
  available_accounts: AITradingAccountOption[]
}

export interface TradeHistoryItem {
  trade_id: string
  user_input: string
  action: string
  symbol: string | null
  quantity: number | null
  price: number | null
  status: string
  confidence: number | null
  risk_level: string | null
  ai_reasoning: string | null
  dry_run: boolean
  created_at: string | null
  executed_at: string | null
}

/**
 * Send a natural language trading instruction.
 */
export async function executeTrade(params: {
  message: string
  gateway_id?: string
  account_id?: string
  dry_run?: boolean
  auto_confirm?: boolean
}): Promise<AITradingResponse> {
  return request.post<AITradingResponse>(`${API_BASE}/execute`, {
    message: params.message,
    gateway_id: params.gateway_id || null,
    account_id: params.account_id || null,
    dry_run: params.dry_run ?? true,
    auto_confirm: params.auto_confirm ?? false,
  })
}

/**
 * Confirm or reject a pending trade.
 */
export async function confirmTrade(params: {
  trade_id: string
  confirmed: boolean
  user_note?: string
}): Promise<{ trade_id: string; status: string; message: string }> {
  return request.post(`${API_BASE}/confirm`, params)
}

/**
 * Get AI trading configuration.
 */
export async function getTradingConfig(): Promise<AITradingConfig> {
  return request.get<AITradingConfig>(`${API_BASE}/config`)
}

/**
 * Get AI trading history.
 */
export async function getTradingHistory(limit = 20): Promise<{
  total: number
  items: TradeHistoryItem[]
}> {
  return request.get(`${API_BASE}/history`, { params: { limit } })
}


/**
 * Generate AI reflection on a completed trade.
 */
export async function reflectOnTrade(tradeId: string): Promise<{
  success: boolean
  trade_id?: string
  reflection?: string
  message?: string
}> {
  return request.post(`${API_BASE}/reflect/${tradeId}`)
}

File: src/backend/app/schemas/ai_trading.py
"""Schemas for AI-driven natural language trading.

Defines the data structures for trading intents, execution requests,
risk assessments, and trading logs.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class TradeAction(str, Enum):
    """Supported trading actions."""

    BUY = "buy"
    SELL = "sell"
    CLOSE = "close"
    CANCEL = "cancel"
    QUERY = "query"
    MODIFY = "modify"


class OrderType(str, Enum):
    """Order type for trade execution."""

    MARKET = "market"
    LIMIT = "limit"
    STOP = "stop"
    STOP_LIMIT = "stop_limit"


class RiskLevel(str, Enum):
    """Risk assessment level."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class TradeStatus(str, Enum):
    """Status of a trade execution."""

    PENDING_CONFIRMATION = "pending_confirmation"
    CONFIRMED = "confirmed"
    EXECUTING = "executing"
    FILLED = "filled"
    PARTIALLY_FILLED = "partially_filled"
    REJECTED = "rejected"
    CANCELLED = "cancelled"
    FAILED = "failed"


class TradingIntent(BaseModel):
    """Structured trading intent parsed from natural language."""

    action: TradeAction
    symbol: str | None = None
    exchange: str | None = None
    quantity: float | None = None
    price: float | None = None
    order_type: OrderType = OrderType.MARKET
    stop_loss: float | None = None
    take_profit: float | None = None
    reason: str = ""
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    risk_level: RiskLevel = RiskLevel.MEDIUM
    raw_input: str = ""
    additional_params: dict[str, Any] = Field(default_factory=dict)


class RiskAssessment(BaseModel):
    """Risk assessment result from the risk guard."""

    approved: bool
    risk_level: RiskLevel
    warnings: list[str] = Field(default_factory=list)
    blocked_reasons: list[str] = Field(default_factory=list)
    requires_confirmation: bool = False
    max_loss_estimate: float | None = None
    position_impact: str | None = None


class AITradingRequest(BaseModel):
    """Request to execute a natural language trade."""

    message: str = Field(..., min_length=1, max_length=500)
    gateway_id: str | None = None
    account_id: str | None = None
    dry_run: bool = True
    auto_confirm: bool = False
    knowledge_base_id: str | None = None
    conversation_id: str | None = None


class AITradingResponse(BaseModel):
    """Response from AI trading execution."""

    trade_id: str
    intent: TradingIntent
    risk_assessment: RiskAssessment
    status: TradeStatus
    message: str
    execution_result: dict[str, Any] | None = None
    ai_reasoning: str = ""
    suggestions: list[str] = Field(default_factory=list)
    requires_confirmation: bool = False
    degraded: bool = False
    diagnostic_message: str | None = None


class TradeConfirmRequest(BaseModel):
    """Request to confirm a pending trade."""

    trade_id: str
    confirmed: bool
    user_note: str | None = None


class TradeConfirmResponse(BaseModel):
    """Response after trade confirmation."""

    trade_id: str
    status: TradeStatus
    message: str
    execution_result: dict[str, Any] | None = None


class AITradingHistoryItem(BaseModel):
    """A single AI trading history entry."""

    trade_id: str
    user_input: str
    intent: TradingIntent
    risk_assessment: RiskAssessment
    status: TradeStatus
    execution_result: dict[str, Any] | None = None
    ai_reasoning: str = ""
    reflection: str | None = None
    created_at: datetime
    executed_at: datetime | None = None


class AITradingHistoryResponse(BaseModel):
    """Response for trading history query."""

    total: int
    items: list[AITradingHistoryItem]


class AITradingConfigResponse(BaseModel):
    """Current AI trading configuration."""

    enabled: bool
    default_mode: str = "paper"
    max_single_trade_amount: float = 10000.0
    max_daily_trades: int = 50
    max_position_ratio: float = 0.3
    require_confirmation_above: float = 5000.0
    blocked_symbols: list[str] = Field(default_factory=list)
    available_gateways: list[dict[str, Any]] = Field(default_factory=list)
    available_accounts: list[dict[str, Any]] = Field(default_factory=list)


class ConditionalOrderCreate(BaseModel):
    """Request to create a conditional (trigger) order."""

    condition: str = Field(..., min_length=1, max_length=500)
    action_message: str = Field(..., min_length=1, max_length=500)
    gateway_id: str | None = None
    dry_run: bool = True
    expiry_hours: float = 24.0


class ConditionalOrder(BaseModel):
    """A conditional order waiting to be triggered."""

    id: str
    user_id: str
    condition: str
    action_message: str
    gateway_id: str | None = None
    dry_run: bool = True
    status: str = "active"  # active, triggered, expired, cancelled
    created_at: str
    expires_at: str
    triggered_at: str | None = None


class ConditionalOrderListResponse(BaseModel):
    """Response for listing conditional orders."""

    total: int
    items: list[ConditionalOrder]
